### ISSUE TITLE
Which lines can contain extraneous data?

### ISSUE THREAD

**Issue by LegionMammal978** [pdf-association/pdf-issues#272]
**Describe the bug**
Occasionally, the standard says that a line of text must contain some particular keyword or data. However, its wording is often unclear on whether that same line can contain other data. First, in 7.5.2:

> If a PDF file contains binary data, as most do (see 7.2, "Lexical conventions"), the header line shall be immediately followed by a comment line containing at least four binary characters—that is, characters whose codes are 128 or greater.

Can a "comment line" contain arbitrary data before the percent sign? If so, can any of the four binary characters be placed before the percent sign, and is the data before the percent sign interpreted as part of the file body? If not, can any white-space be placed before the percent sign? This should be clarified.

In 7.5.4:

> Each cross-reference section shall begin with a line containing the keyword **xref**.

Can this line contain any data before or after the `xref`? If not, this should be clarified as "containing only the keyword".

In 7.5.5:

> The two preceding lines shall contain, one per line and in order, the keyword **startxref** and the byte offset from the beginning of the PDF file to the beginning of the **xref** keyword in the last cross-reference section or the beginning of the previous cross-reference stream.

Can these two lines contain any data before or after the `startxref` or byte offset? If not, this should be clarified as "shall contain only".

**Comment by lrosenthol** [comment]
As defined in 7.2.4, a comment is everything after the `%`.   So while not explicitly defined, a "comment line" would be a line that starts with a `%`.   @petervwyatt definitely worth adding that definition or clarifying.

On the xref sections and startxref, any data preceding those keywords would be ignored - so, sure, in theory you could have other stuff on that line...

**Comment by LegionMammal978** [comment]
> On the xref sections and startxref, any data preceding those keywords would be ignored - so, sure, in theory you could have other stuff on that line...

The problem with that is that it would be impossible in the general case for a PDF reader to correctly interpret, e.g.:

```
startxref
The byte offset is 123. However, this line also contains the numbers 456 and 789.
%%EOF
```

(Although if the wording is changed, it might be worth clarifying whether the `startxref` and byte offset lines can end in a comment.)

**Comment by petervwyatt** [comment]
Subclause 7.2.4 Comments is very clear that comments start with a `%` and go until (but not including!) the EOL - they are not "lines" per-se in that they have to be preceded with a starting EOL:
> Any occurrence of the PERCENT SIGN (25h) outside a string or inside a content stream (see 7.8.2, "Content streams") introduces a comment. The comment consists of all characters after the PERCENT SIGN and up to but not including the end-of-the-line marker. A PDF processor should ignore comments. PDF processors shall treat comments as single white-space characters for the purposes of lexical conversion (see 7.2, "Lexical conventions"). That is, a comment separates the token preceding it from the one following it.

I believe the confusion is from the use of the phrase "comment line" since that might be interpreted to infer a preceding EOL. However in the specific case for the binary marker comment line this is a **correct inference**:

* The phrase "_comment line_" only occurs once in PDF 2.0 in subclause 7.5.2 File Header:
> If a PDF file contains binary data, as most do (see 7.2, "Lexical conventions"), the header line shall be immediately followed by a comment line containing at least four binary characters–that is, characters whose codes are 128 or greater.

* Subclause 7.5.2 File header also enforces that the file header ends with an EOL so it is indeed a "file header line":
> The file header shall consist of `%PDF–1.n` or `%PDF–2.n` followed by a single EOL marker, where ‘n’ is a single digit number between 0 (30h) and 9 (39h).

The requirement also explicitly states "immediately" so nothing can precede it - altho' this should possibly allow for whitespace. Together this means the header + binary comment line of a compliant PDF needs to look like this:
`%PDF-x.y<EOL>%<b1><b2><b3><b4> ... <EOL>`

I think the only improvement to the text in 7.5.2 would be to use the correct term "file header" and possibly also repeat the "... followed by a single EOL marker" - otherwise the text is technically correct.

As to the more generic problem of the word "line" in describing PDF syntax - that is more problematic and is already covered by Issue #112.

**Comment by LegionMammal978** [comment]
> The requirement also explicitly states "immediately" so nothing can precede it - altho' this should possibly allow for whitespace.

The *comment line* is what must immediately follow the header line. This does not necessarily imply that the *comment* contained within the *comment line* must immediately follow the header line. The intuitive reading is that a *comment line* contains a *comment* and nothing other than that *comment*, but nothing defines this explicitly.

**Comment by petervwyatt** [comment]
> This does not necessarily imply that the comment contained within the comment line must immediately follow the header line.

"_Comment line_" is not a formal T&D nor is it an informal term defined anywhere (in italics) - in fact this phrase is used only once in the entire PDF spec. 

However the definition of a "comment" **IS** formally defined in 7.2.4 with the statement "Any occurrence of the PERCENT SIGN (25h) outside a string or inside a content stream (see 7.8.2, "Content streams") introduces a comment.". This clearly states that the `%` starts a comment. Furthermore, the formal definition of a comment **excludes** the EOL so there is no such thing as a "comment line".

So your example is not a "comment line" but a "line including a comment" and thus is not "immediately following".

If we really wanted to get this pedantic we might change to be:
> If a PDF file contains binary data, as most do (see 7.2, "Lexical conventions"), the file header line shall be immediately followed by a line containing only a comment that starts with at least four binary characters immediately after the PERCENT SIGN – that is, characters whose codes are 128 or greater.
